emu/src/bitwise.rs

```diff
@@ -55,6 +55,50 @@ where
         }
     }
 
+    fn set_bits(&self, bits_range: RangeInclusive<u8>, value: Self) -> Self {
+        let start = bits_range.start();
+        let length = bits_range.len() as u32;
+        let self_bits: u128 = self.clone().into();
+
+        // Set all of the desider bits to 1 then & it with the value provided
+        // considering the right amount of bits (given by bits_range.len()).
+        // Order goes from lsb to msb (right to left).
+        let mask = (2_u128.pow(length)) - 1;
+        let value_bits: u128 = value.into();
+        let value_bits: u128 = (value_bits & mask) << start;
+
+        // Now, shift the mask then flip it so we can choose where to insert
+        // our value bits
+        let reverse_mask = !(mask << start);
+
+        // Say we have self being a u16 value with the following bits:
+        //     0b0000....10010011_01110011_u128
+        // and we want to set bits 7..=10, 4 bits starting from 7 to 0b1001.
+        //
+        // The constant `reverse_mask` will look something like:
+        //     0b1111....11111000_01111111_u128
+        // which is helpful for clearing the bits we are about to set.
+        //
+        //     0b0000....10010011_01110011_u128 &
+        //     0b1111....11111000_01111111_u128
+        //     --------------------------------
+        //     0b0000....10010000_01110011_u128
+        //
+        // The value above can be used in bit-or with our value bits
+        // to obtain the expected result:
+        //
+        //     0b0000....10010000_01110011_u128 |
+        //     0b0000....00000100_10000000_u128 =
+        //     --------------------------------
+        //     0b0000....10010100_11110011_u128
+
+        <Self as TryFrom<u128>>::try_from((self_bits & reverse_mask) | value_bits).unwrap()
+    }
+
+    fn set_bits_mut(&mut self, bits_range: RangeInclusive<u8>, value: Self) {
+        self.clone_from(&self.set_bits(bits_range, value))
+    }
```

Comment:
This is the real `set_bit` because it changes the value in the same instance where you call method. The other returns a copy with some bits changed...

Reply:
I would try to be consistent with the existing methods. For example we only have `set_bit`, `set_byte`, etc... which change the instance on which are called, without returning `Self`.

I mean we can have this behaviour but I would introduce it in every function, just to be consistent (would it make the usage better?)

Reply:
Oh ok, I get it now, it makes sense now I guess. Gonna change it np :+1: